### PR TITLE
add support for batch/v1/Jobs in crossplane-extension-manager install.yaml

### PR DESCRIPTION
**Pull request by displague**
This PR teaches the extension-manager how to install Kubernetes Jobs.
Previously, only Deployments could be installed.

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the RBAC permissions in [`clusterrole.yaml`] to include any new types.

[documentation]: https://github.com/crossplaneio/crossplane/tree/master/docs
[examples]: https://github.com/crossplaneio/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplaneio/crossplane/tree/master/PendingReleaseNotes.md
[`clusterrole.yaml`]: https://github.com/crossplaneio/crossplane/blob/master/cluster/charts/crossplane/templates/clusterrole.yaml

### CODE REVIEW

File: pkg/controller/extensions/extension/extension.go [hunk 1]
```diff
@@ -241,6 +246,41 @@ func (h *extensionHandler) processDeployment(ctx context.Context) error {
...
+			OwnerReferences: []metav1.OwnerReference{ref},
+		},
+		Spec: jobSpec,
+	}
+	if err := h.kube.Create(ctx, j); err != nil && !kerrors.IsAlreadyExists(err) {
+		return errors.Wrap(err, "failed to create job")
+	}
+
+	// save a reference to the extension's controller
+	h.ext.Status.ControllerRef = &corev1.ObjectReference{
```
Comment (negz) [inline pkg/controller/extensions/extension/extension.go]: You could also use `meta.ReferenceTo(j, JobGroupVersionKind)` here if you wanted.

File: pkg/extensions/unpack.go [hunk 2]
```diff
@@ -81,7 +84,7 @@ func findRegistryRoot(fs afero.Fs, dir string) string {
 	return dir
 }
 
-func doUnpack(fs afero.Fs, root string) (string, error) {
+func doUnpack(fs afero.Fs, root string) (string, error) { // nolint:gocyclo
```
Comment (negz) [inline pkg/extensions/unpack.go]: I recommend leaving a small comment with a bit of context when overriding the linter, even if it's just something like "this function is more readable with slightly higher cyclomatic complexity"

File: pkg/controller/extensions/extension/extension.go (hunk 1 above)
Comment (displague) [reply pkg/controller/extensions/extension/extension.go]: Is 60dc62f0292f3f6ead6170c81cebed1224164eda what you were thinking?

Comment (negz) [reply pkg/controller/extensions/extension/extension.go]: Does `d.GroupVersionKind()` always return the GVK of a deployment? I'm unfamiliar with that particular method, but I know it's quite common for objects to attempt to return their GVK based on their embedded type metadata, which is often unset.

Comment (displague) [reply pkg/controller/extensions/extension/extension.go]: Good catch.  `TypeMeta` is not explicitly set here,
https://github.com/crossplaneio/crossplane/pull/583/files#diff-0dc4437d46271355bb75e3c9b14295ddR225-R232 .. 
So what response does this cause ellicit from `GroupVersionKind()`? .. `schema.GroupVersionKind{"","",""}`

(I took advantage of #579 to dig into this)

I'm defining the GVK in advance in d7e8d08 and referencing that when setting the controller reference.

Comment (negz) [reply pkg/controller/extensions/extension/extension.go]: You could also lookup the Job or Deployment's GVK in the controller manager's scheme, roughly like:

```go
type extensionHandler struct {
	kube  client.Client
        typer runtime.ObjectTyper  // Satisfied by mgr.GetScheme()
	ext   *v1alpha1.Extension
}

// Assume eh is an &extensionHandler{}
ref := meta.ReferenceTo(j, resource.MustGetKind(j, eh.typer)
```

Comment (displague) [reply pkg/controller/extensions/extension/extension.go]: I'm having a hard time figuring out the least obtrusive way to fit that mgr object into the extensionHandlerFactory `newHandler` method.

What is the benefit of using the `MustGetKind(...typer)` approach compared to using hard coded GKV (especially considering that this follows the creation of the Kubernetes object using typed objects -- `&appv1.Deployment{}`)?

Comment (negz) [reply pkg/controller/extensions/extension/extension.go]: > What is the benefit of using the MustGetKind(_, typer) approach compared to using hard coded GKV

I don't think there's an obvious benefit, so if it's hard to do a hard coded GVK should be fine.

**Comment by jbw976** [comment]
Looks like the SonarCloud analysis failed because code coverage on new code isn't high enough.  The main culprit looks to be `pkg/extensions/unpack.go`, so we should probably add a new test case that exercises the parsing/unmarshal of Jobs, similar to how we are already testing that for Deployments.

**Comment by displague** [comment]
@jbw976 I actually did add that test, https://github.com/crossplaneio/crossplane/pull/583/files#diff-69a52b86d960526dcdfc9a82c3d5a582 but came out 1.2 percent short of the 80% sonarcloud was looking for.

File: pkg/controller/extensions/extension/extension.go [hunk 3]
```diff
@@ -225,18 +236,47 @@ func (h *extensionHandler) processDeployment(ctx context.Context) error {
...
+	gvk := schema.GroupVersionKind{
+		Group:   batch.GroupName,
+		Kind:    "Job",
+		Version: batch.SchemeGroupVersion.Version,
 	}
+	j := &batch.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            controllerJob.Name,
+			Namespace:       h.ext.Namespace,
+			OwnerReferences: []metav1.OwnerReference{ref},
```
[the remark this answers is not in the document]
Comment (displague) [reply pkg/controller/extensions/extension/extension.go]: Yes. The job is removed with the extension.

```console
$ kubectl get pods,jobs,extension,extensionrequests -A
NAMESPACE            NAME                                                     READY   STATUS      RESTARTS   AGE
container-registry   pod/registry-6c99589dc-6lvf7                             1/1     Running     6          7h33m
crossplane-system    pod/crossplane-745f86b9bb-vdbbm                          1/1     Running     0          80s
crossplane-system    pod/crossplane-extension-manager-f599d94d5-nfrn6         1/1     Running     0          80s
default              pod/default-http-backend-5d5ff5d4f5-xkjrn                1/1     Running     6          7h33m
default              pod/nginx-ingress-microk8s-controller-8r29c              1/1     Running     6          7h33m
default              pod/sample-extension-metacontroller-from-package-bzhrj   0/1     Completed   0          37s
kube-system          pod/hostpath-provisioner-65cfd8595b-5sm85                1/1     Running     6          7h33m
kube-system          pod/tiller-deploy-64f5b9869-cmzvp                        1/1     Running     5          6h18m

NAMESPACE   NAME                                                     COMPLETIONS   DURATION   AGE
default     job.batch/sample-extension-metacontroller-from-package   1/1           4s         37s

NAMESPACE   NAME                                                                              READY   VERSION   AGE
default     extension.extensions.crossplane.io/sample-extension-metacontroller-from-package   False   0.0.1     27s

NAMESPACE   NAME                                                                                     READY   SOURCE   PACKAGE                                     CRD   AGE
default     extensionrequest.extensions.crossplane.io/sample-extension-metacontroller-from-package   True             displague/sample-extension-metacontroller         38s

$ kubectl delete -f install-the-extension.yaml 
extensionrequest.extensions.crossplane.io "sample-extension-metacontroller-from-package" deleted

$ kubectl get pods,jobs,extension,extensionrequests -A
NAMESPACE            NAME                                               READY   STATUS    RESTARTS   AGE
container-registry   pod/registry-6c99589dc-6lvf7                       1/1     Running   6          7h35m
crossplane-system    pod/crossplane-745f86b9bb-vdbbm                    1/1     Running   0          2m54s
crossplane-system    pod/crossplane-extension-manager-f599d94d5-nfrn6   1/1     Running   0          2m54s
default              pod/default-http-backend-5d5ff5d4f5-xkjrn          1/1     Running   6          7h35m
default              pod/nginx-ingress-microk8s-controller-8r29c        1/1     Running   6          7h35m
kube-system          pod/hostpath-provisioner-65cfd8595b-5sm85          1/1     Running   6          7h35m
kube-system          pod/tiller-deploy-64f5b9869-cmzvp                  1/1     Running   5          6h20m
```

**Comment by displague** [comment]
Coverage met and tests updated in #653baf4

**Pull request merged**